documentation/design/proposals/authn_authz_for_1.0.adoc

```diff
@@ -0,0 +1,329 @@
+availability and performance.  It will be implemented in Go.
+
+The Access Control Server will be created by EnMasse Operator according to configuration within the MessagingInfra object.
+
+The Access Control Server will provide an AMQP SASL endpoint that implements `authServicePlugin` interface.
+
+On each incoming AMQP connection, the Access Control Server will use TLS SNI will allow it to determine which
+`AccessControlService` to apply when processing each authentication event.
+
+The algorithm for processing each authentication event:
+
+1. Router connects to the AMQP SASL endpoint associated with the router listener.
+1. Access Control Server uses the SNI information to resolve the applicable `AccessControlService`
+1. Access Control Service uses the configuration of the `AccessControlService` determine the avaiable SASL mechanism(s).
+1. SASL-MECHANISMS performative flows the Client.
+1. Client selects mechanism
+1. SASL-INIT flows to the Access Control Server
+1. SASL-CHALLENGE / SASL RESPONSE as necessary
+1. Access Control Server calls out to the external authentication service / uses `MessagingUser` / `TokenReview` (as appropiate) to determine if the authentication is successful.
+1. SASL-OUTCOME flows to the Client.
+1. Access Control Server uses the `MessagingRole`, and `MessagingRoleBinding` to calculate the `address-authz`.
+1. AMQP OPEN flows to the Router carrying the authorisation policy.
+
+===== Provider Behaviour
+
+====== External LDAP Provider
+
+Allows the use of an Open Directory implementing the LDAP protocol to provide an identity and, optionally, group information.
+The Access Control Server will perform a simple bind against the LDAP in order to determine the success or failure of the authentication.
+
+Simple bind requires the plain text password from the client, so SASL PLAIN will be offered when this identity provider is used.
+
+Later, support will be added for LDAP SASL.
+
+====== External AMQP SASL Provider
+
+Allows the use of an AMQP SASL authentication endpoint that implements the `authServicePlugin` protocol.  When this identity
+provider is used, the Access Control Server acts a SASL proxy.  It will connect to the external provider and relay the
+the AMQP SASL performatives to/from the client.
+
+The response from the External AMQP SASL server must provides an `authenticated-identity`.  It may also provide
+`groups` and may provide `address-authz`.  If the external server provides an `address-authz`, it will be ANDed with the `address-authz` calculated from the `MessagingRoles` and `MessagingRoleBindings`.
+
+The SASL mechanisms offered when this identity provider are in use depend on the mechansims required by the External AMQP SASL server.
+
+====== Internal Identity Provider
```

Comment:
Would a better name be Kubernetes Identity Provider, Local Identity Provider or Cluster Identity Provider?

Reply:
Yeah, I agree Internal is not a great name.   *Kubernetes Identity Provider* might make user think that it somehow uses a user concept from Kubernetes.  Cluster Identity Provider might have the user believe it is somehow cluster scoped.  Local Identity Provider works I think, or maybe NamespaceIdentityProvider?

Reply:
I went with NamespaceIdentityProvider.